TextEdit respects content margin from StyleBox

pouleyKetchoupp opened the pull request:
`TextEdit` adjusts x & y offset according to the corresponding `StyleBox` when in normal or read-only mode.

The x offset from normal style is already applied in any case so it has to be substracted when calculating the x offset in this part of code.

Now `TextEdit` acts the exact same way as `LineEdit` when it comes to content margins.

Fixes #32867

pouleyKetchoupp committed:
TextEdit respects content margin from StyleBox

Now TextEdit adjusts x & y offset according to the corresponding
StyleBox when in normal or read-only mode.

In order to handle bottom content margin, wrapped lines that are entirely outside the stylebox content area are not drawn.

pouleyKetchoupp commented:
@RandomShaper Could you please check again? I've just added a change to draw only visible lines, in order to respect the bottom margin as well. before it was drawing the text for wrapped lines as well even if not visible.

Paulb23 commented:
@pouleyKetchoupp `TextEdit` needs to use `draw_amount` as when `smooth_scroll` is enabled, we could be half way down a line, in this case we need to draw the top half of the next line. 

Secondly, adding the wrapped rows for the first line prevents the below from happening:
![visible_lines](https://user-images.githubusercontent.com/6584330/107570727-d6c54c00-6be1-11eb-8c13-04f67d7a0574.gif)

pouleyKetchoupp commented:
@Paulb23 Ok, makes sense! I'll find a better way to clip the bottom part when there's a bottom content margin then.

pouleyKetchoupp commented:
I've reverted my previous change and instead, wrapped lines that are entirely outside the content margin of the stylebox are not drawn, so it still displays partial lines in the cases it did before.

I've also removed the case that adapts x/y offsets to the focused stylebox when focused, because I noticed some discrepancies with mouse clicks when margins are different from the normal style, and changing the whole class to take the focus stylebox into account is a bit too involved for this PR. So for now, it still relies on normal stylebox when focused and ignores specific content margins from the focus stylebox.

Paulb23 commented on scene/gui/text_edit.cpp:
```diff
@@ -1019,17 +1019,34 @@ void TextEdit::_notification(int p_what) {
+						}
+
+						if (ofs_y - row_height > max_size_y) {
+							// Line is outside the content margin, clip.
```
Would we need to set `i = draw_amount;` to also break from the outer loop?

pouleyKetchoupp replied:
Good catch, thanks! I've added that so it can break earlier when there's nothing else to draw.
Actually, I've also removed the condition `if (line_wrap_index != 0)` so it can apply the clipping on lines after line breaks too.

pouleyKetchoupp commented:
In addition to the change I've made from your comment for bottom margin clipping, I've made changes to handle top margin clipping properly.

It's useful when scrolling in the TextEdit in case there's top content margin, and in this case I can't just skip the line ( it would set the wrong color on following lines in some cases), so I'm setting a `clipped` variable which is used to avoid drawing the characters, caret and other features.

One last thing is I've removed a line at the very end of the line loop: `ofs_y += ldata->get_line_descent(line_wrap_index);`
It does nothing currently as ofs_y is set on each iteration, but I'd like to make sure it's not supposed to be used for something and would need to be fixed instead.

Paulb23 commented on scene/gui/text_edit.cpp:
```diff
@@ -1019,17 +1019,43 @@ void TextEdit::_notification(int p_what) {
 					ofs_y -= cursor.wrap_ofs * row_height;
 					ofs_y -= get_v_scroll_offset() * row_height;
 
+					int top_limit_y = 0;
+					int bottom_limit_y = get_size().height;
+					if (readonly) {
+						top_limit_y += cache.style_readonly->get_margin(SIDE_TOP);
+						bottom_limit_y -= cache.style_readonly->get_margin(SIDE_BOTTOM);
+					} else {
+						top_limit_y += cache.style_normal->get_margin(SIDE_TOP);
+						bottom_limit_y -= cache.style_normal->get_margin(SIDE_BOTTOM);
+					}
```
Might be worth pre-calculating this outside of the loops.

pouleyKetchoupp replied:
Done and pushed.

Paulb23 reviewed (commented):
Minor nitpick made inline, otherwise looks good to me.

akien-mga commented:
Thanks!

akien-mga commented:
It's not trivial to cherry-pick, so a dedicated PR would be welcome for `3.2`.

pouleyKetchoupp commented:
@akien-mga Dedicated 3.2 PR here: #46060

ondesic commented:
As of 3.4.1 Stable, Content Margins of StyleBoxes have been unpredictable in each Node.  

Here are two images.  Both are a TextEdit node with a StyleBoxFlat.  The first image is NO Content Margin. The second image is a Top Content Margin of 20.  These are both in the Editor.
![3](https://user-images.githubusercontent.com/26915986/146642055-8ba3e258-91fd-4645-844d-722bb332c68c.png)
![4](https://user-images.githubusercontent.com/26915986/146642057-bc0ff68d-3f22-4988-b251-22771d2fdb4e.png)

The Editor shows a correct applying of Top Content Margin (however, notice that the scrollbar of the second image has shortened both on top AND bottom).  
Now, when the app is run, here is how it looks:
![5](https://user-images.githubusercontent.com/26915986/146642129-c61a661c-2cf5-4518-9aef-1913832b02f4.png)

I love that the top margin is actually working now.  It used to not work.  However, notice that even though I only set the Top Content Margin, it has been applied to both Top AND Bottom, shrinking it on both ends.  This isn't how it should be.  This should only happen if the Top AND Bottom Content Margin are both set to 20.

Currently, setting the Bottom Content Margin is only changing the scrollbar height, but not the actual content.

KoBeWi commented:
@ondesic Open a new issue if it wasn't already reported.

ondesic commented:
I felt that the title fell under the problem I was noticing with TextEdits, content Margins and Styleboxes.  I am happy to start a new issue, however, 50% of the time when I do, it is closed as a duplicate.  Just trying to avoid that.  I'll create a new issue.
